<div style="padding:20px;background-color:papayawhip;" > 
<h3 style="color:chocolate"> <i class="fa fa-info" aria-hidden="true"> </i> &nbsp; Remarque introductive &nbsp;  <i class="fa fa-info" aria-hidden="true"></h3> 
<p> Ce fichier est fait pour être lu sur le site <a href="https://notebook.basthon.fr/"><img src='https://notebook.basthon.fr/assets/efede5218c9087496f16.png' style="border: 0; display:inline; margin: 0 5px; height:30px" alt="Basthon"/></a>. C'est ce logo qui doit s'afficher en haut à gauche de cette fenêtre. <br>
    
Si vous l'avez ouvert avec un autre programme, comme Jupyter notebook, vous riquez de rencontrer quelques bugs. <br>
Veuillez cliquez sur <a href="https://notebook.basthon.fr/">ce lien</a> et y charger ce fichier à l'aide du bouton "Ouvrir"  &nbsp; <i class="fa fa-folder" aria-hidden="true"> </i>
</p> </div>    

    
!!! warning Toujours sauvegarder votre travail

- Si c'est la première fois que vous ouvrez la leçon, alors commencer par enregistrer ce notebook sous le nom `Lecon1b_print.ipynb` dans vos documents sous le dossier indiqué par votre enseignant-e (utiliser le bouton <i class="fa fa-save"> </i>) ou exercice. 
- Avant de fermer la fenêtre il ne faut pas oublier de réenregistrer le notebook avec le même nom (répondre oui à "remplacer le fichier").
- Pour reprendre votre travail, il suffit d'ouvrir le fichier .ipynb en cliquant sur le bouton "Ouvrir"  &nbsp; <i class="fa fa-folder" aria-hidden="true"> 
!!!


# 1b. Afficher de l'information et faire des calculs
Dans ce notebook nous apporterons quelques compléments aux notions abordées dans le premier *notebook*.

Les programmes ne se limitent pas à la réalisation de graphiques ou de dessin. En fait, pendant longtemps, le seul moyen d'interagir avec un programme était par le biais du texte.

Par conséquent, les programmes sont également capables de manipuler du texte, y compris de lire, d'écrire et d'afficher des textes sur un écran. Dans ce contexte, les commandes `from turtle import *` et `done()`, habituellement utilisées pour le dessin avec la tortue, deviennent inutiles. En terme informatique, les textes sont appelés des chaînes de caractères (str en python).

## L'instruction Print
La première instruction que nous allons voir est la fonction print qui permet d'afficher du texte à l'écran.

En Python, l'instruction `print("blabla")` affiche le texte `blabla` dans la console (partie nommée Sortie). 


!!! danger Important

les parenthèses et les guillemets sont nécessaires. On verra pourquoi par la suite.
!!!

!!! note &nbsp; Exercice 1

1. Tester le programme suivant en appuyant sur Ctrl + Enter ou le bouton Exécuter en haut.
2. Remplacez le texte blabla blablabla par le titre de votre film préféré.

!!!


In [ ]:
print("blabla blablabla")

!!! note &nbsp; Exercice 2

Écrivez un programme qui affiche exactement :

`Vive l'instruction print !`

!!!


In [ ]:
# à compléter

### Afficher plusieurs lignes de texte
Pour afficher plusieurs lignes, on peut utiliser plusieurs appels à print(...) :


```python
print("Première ligne")
print("Deuxième ligne")
```

!!! note &nbsp; Exercice 3a

Compléter le programme pour afficher le texte sur trois lignes comme ci-dessous:
```
Bonjour !
Comment allez-vous ?
Il fait beau aujourd'hui.
```

!!!


In [ ]:
print("Bonjour !")
# À compléter

### Repérer et corriger des erreurs


!!! note &nbsp; Exercice 3b

Le programme suivant contient plusieurs erreurs de syntaxe :

!!!


In [ ]:
print"Aujourd'hui"
print(le temps est mitigé,)
print("mais je vais quand même sortir.)

### Des programmes qui calculent

La fonction `print` ne sert pas uniquement à afficher du texte, elle permet également d'afficher le résultat de l'évaluation d'expressions, notamment des opérations arithmétiques.

Par exemple, `print(3+3)` affichera le résultat de l'opération 3+3, soit... `6`. 

!!! warning Afficher une chaîne de caractères Vs afficher le résultat d'une expression

Pour effectuer le calcul, il faut ôter les guillemets, sinon le programme affichera le texte de l'expression et non son résultat. Par exemple, l'instruction `print("3+3")` afficherait simplement `3+3` et pas `6`.
!!!

!!! note &nbsp; Exercice 4

Teste le code ci-dessous puis modifiez-le afin d'afficher le résultat de **18975 + 98756** dans la console, et vérifiez en appuyant sur Ctrl + Enter.

!!!


In [ ]:
print(3+3)

### 🔢 Les opérations avec les nombres en Python

En programmation **Python**, on dispose des opérateurs mathématiques suivants :  

| Opérateur | Signification              | Exemple | Résultat |
|-----------|-----------------------------|---------|----------|
| `+`       | Addition                   | `5 + 3` | `8` |
| `-`       | Soustraction                | `9 - 4` | `5` |
| `*`       | Multiplication              | `6 * 7` | `42` |
| `/`       | Division (résultat flottant)| `9 / 2` | `4.5` |
| `**`      | Puissance                   | `2 ** 3` | `8` |
| `//`      | Division entière (quotient) | `9 // 2` | `4` |
| `%`       | Reste de la division entière| `17 % 5` | `2` |

---

📌 **Exemple illustratif** :  

```python
print(9 // 2)   # donnera 4, c'est le quotient de la division de 9 par 2.
print(17 % 5)   # donnera 2, c'est le reste de la division de 17 par 5.
```
![Image de losange](https://github.com/College-Sismondi-Informatique/exercices-1in/blob/main/Notebooks/imgs_chap1/div_eucl.png?raw=true)


!!! note &nbsp; Exercice 5a

Tester tous les opérateurs mathématiques vus ci-dessus (`+`, `-`, `*`, `/`, `**`, `//`, `%`).

!!!


In [ ]:
# Exercice 5a - mon programme


!!! note &nbsp; Exercice 5b

Écrire un programme Python qui calcule :  

$$
\frac{264 \times 53}{22} + 274 
$$

✅ Si votre programme fonctionne correctement, il doit afficher **910**.  

---

💡 **Indice :**  
- Attention à bien utiliser les parenthèses pour respecter l’ordre des opérations.

!!!


In [ ]:
# Exercice 5b - mon programme


### 🖨️ Afficher plusieurs valeurs avec `print`

En Python, il est possible d’afficher plusieurs valeurs sur une **même ligne** avec la fonction `print(...)`.  
Pour cela, il suffit de **séparer les éléments par des virgules**.  

📌 **Exemple :**

```python
print("Aujourd'hui, j'ai", 20 * 2 + 1, "ans")
```
Affichera:
`Aujourd'hui, j'ai 41 ans`.


!!! note &nbsp; Exercice 5c

Reprendre l'exercice précédent pour afficher :
`Le résultat de (264×53)/22+274 est ...` où à la place des `...` il y a le résultat du calcul

!!!


In [ ]:
# Exercice 5c - mon programme


!!! note &nbsp; Exercice 6

Calculez de tête et noter ci-dessous ce qui va être affiché par chacune des instructions suivantes:

    1. print(3+5)
    2. print(7*2)
    3. print(8/4)
    4. print(7/2)
    5. print(7//2)
    6. print(2**3)
    7. print(13%5)
 puis vérifier en cliquant sur `Réponses`.
 
 Le but de l'exercice est d'apprendre l'ensemble des opérateurs en python.

!!!


<details>
<summary style="border-left:3px solid #3c763d; border-radius:2pt; width:100%; color:#3c763d; padding:6px; background-color: #dff0d8"> 
Réponses
</summary>  

<div style="border-left:3px solid #3c763d; border-radius:2pt; color:#3c763d; padding:6px; background-color: #eff0e8">

1. `8`<br>
2. `14`<br>
3. `2`<br>
4. `3.5`<br>
5. `3`<br>
6. `8`<br>
7. `3`<br>
    
</div>
</details>

!!! note &nbsp; Exercice 7

Quelle opération pourrait-on utiliser pour reconnaître si un nombre est pair ou impair? (une opération qui donnerait toujours le même résultat si le nombre est pair et toujours le même résultat si le nombre est impair)

!!!


<details>
<summary style="border-left:3px solid #3c763d; border-radius:2pt; width:100%; color:#3c763d; padding:6px; background-color: #dff0d8"> 
Réponse
</summary>  

<div style="border-left:3px solid #3c763d; border-radius:2pt; color:#3c763d; padding:6px; background-color: #eff0e8">

On utilise le reste de la division entière (opérateur `%` par 2. Si le résultat est 0 le nombre est pair, si il vaut 1 c'est qu'il est impair. 
    
</div>
</details>

!!! note &nbsp; Exercice 8

Calculez de tête et noter ci-dessous ce qui va être affiché par chacune des instructions suivantes:

    1. print(98145 % 2)
    2. print(7841592 % 2)
    3. print(879 % 10)
 puis vérifier en cliquant sur `Réponses`

!!!


<details>
<summary style="border-left:3px solid #3c763d; border-radius:2pt; width:100%; color:#3c763d; padding:6px; background-color: #dff0d8"> 
Réponses
</summary>  

<div style="border-left:3px solid #3c763d; border-radius:2pt; color:#3c763d; padding:6px; background-color: #eff0e8">

1. `1`<br>
2. `0`<br>
3. `9`<br>

    
</div>
</details>

!!! note &nbsp; Exercice 9

Écrivez un programme permettant de vérifier si 23'528'232 est un multiple de 27.

Indice : si un entier est divisible par 27, quel doit être le reste de sa division par 27 ?


!!!


In [ ]:
# Exercice 9
# Mon programme




## Utiliser le binaire avec Python

### Conversion binaire → décimal

En Python, il est possible d’écrire directement un nombre entier en **binaire**.  
Pour cela, on utilise le préfixe **`0b`** devant le nombre.  

---

📌 **Exemple :**

Python accepte l'écriture d'un entier en base 2 avec le préfixe 0b. Par exemple, le nombre binaire `10111` peut être écrit en Python sous la forme :  

```python
print(0b10111)


In [1]:
print(0b10111)

23


!!! note &nbsp; Exercice 10

Écrire un programme Python qui affiche la valeur **décimale** des nombres binaires suivants : 

- $1001001 _{(2)}$

- $111001 _{(2)}$

- $11001001001 _{(2)}$

!!!


In [ ]:
# Exercice 10


---
### Conversion décimal →  binaire
Pour afficher la représentation **binaire** d’un nombre entier en Python,  
on peut utiliser la fonction intégrée `bin(...)`:

In [ ]:
print(bin(23))

!!! note &nbsp; Exercice 11

Écrire un programme Python qui affiche la représentation **binaire** des nombres suivants :

- $14646841 _{(10)}$
- $8156468435 _{(10)}$
- $99999999 _{(10)}$

!!!


In [ ]:
# Exercice 11



## Le cas des nombres à virgule flottante*

Dans ce cours, nous utilisons principalement deux types numériques Python :

- `int` pour les entiers : `3`, `100`, `-20` ;
- `float` pour les nombres à virgule : `3.5`, `-0.001`, `2.0`.

C'est pourquoi les nombres peuvent être de types `int` ou `float`, comme on le voit en utilsant la fontion type:

In [ ]:
print(type(17))
print(type(3.14))

!!! warning En Python, le séparateur décimal est le point, par exemple 4.5

!!!

## Une expérience surprenante

Avant d'exécuter la cellule suivante, prédisez le résultat de :

\[
0.1 + 0.2
\]

In [2]:
print(0.1 + 0.2)

0.30000000000000004


On obtient généralement :

0.30000000000000004

Ce n'est pas parce que Python « ne sait pas additionner ».

Le problème vient de la représentation : de la même manière que 1/3 donne un développement décimal infini 0.3333…, le nombre décimal 0.1 possède un développement binaire infini. L'ordinateur doit donc stocker une approximation sur un nombre fini de bits. Les nombres à virgules  `float` sont représentés en utilisant une décomposition parfois approximative en puissance de 2 (voir la théorie des nombres à virgules flottantes).

Explication détaillée: <a href="https://cours-info-nsi-56cb21fb8d50d1af289d9150b1042c010be13879c6badb00.forge.apps.education.fr/premiere/Donnees/c4c_nb_flottants/" target="_blank">lien</a>

??? tip Pour aller plus loin
Si on représente le nombre $0,1_{(10)}$ avec un codage binaire à virgule flottante on obtient un deéveloppement périodique:
$$0,1_{(10)} = 0,00011001100110011001100|\overline{1100}_{(2)}$$
Par exemple avec 24 bits, l'erreur commise est donc d'environ 
$$0,000000000000000000000001100 \simeq 9,54 \times 10^{-8}$$
???


??? tip Problèmes d'arrondi: Missile guerre du Golf (1991) - Fusée Ariane 5 (1996)

Voici deux exemples de désastres causés par une mauvaise gestion des erreurs d'arrondi :

Le 25 février 1991, pendant la Guerre du Golfe, une batterie américaine de missiles Patriot, à Dharan (Arabie Saoudite), a échoué dans l'interception d'un missile Scud irakien. Le Scud a frappé un baraquement de l'armée américaine et a tué 28 soldats. La commission d'enquête a conclu à un calcul incorrect du temps de parcours, dû à un problème d'arrondi. Les nombres étaient représentés en virgule fixe sur 24 bits. Le temps était compté par l'horloge interne du système en 1/10 de seconde. Malheureusement, 1/10 n'a pas d'écriture finie dans le système binaire : 1/10 = 0,1 (dans le système décimal) = 0,0001100110011001100110011... (dans le système binaire). L'ordinateur de bord arrondissait 1/10 à 24 chiffres, d'où une petite erreur dans le décompte du temps pour chaque 1/10 de seconde. Au moment de l'attaque, la batterie de missile Patriot était allumée depuis environ 100 heures, ce qui avait entraîné une accumulation des erreurs d'arrondi de 0,34 s. Pendant ce temps, un missile Scud parcourt environ 500 m, ce qui explique que le Patriot soit passé à côté de sa cible.

Le 4 juin 1996, une fusée Ariane 5 a explosé 40 secondes après l'allumage. La fusée et son chargement avaient coûté 500 millions de dollars. La commission d'enquête a rendu son rapport au bout de deux semaines. Il s'agissait d'une erreur de programmation dans le système inertiel de référence. À un moment donné, un nombre codé en virgule flottante sur 64 bits (qui représentait la vitesse horizontale de la fusée par rapport à la plate-forme de tir) était converti en un entier sur 16 bits. Malheureusement, le nombre en question était plus grand que 32768, le plus grand entier que l'on peut coder sur 16 bits, et la conversion a été incorrecte.
???

!!! note &nbsp; Exercice 12

Écrivez un programme qui affiche une phrase de la forme :

```text
Dans une image codée en 24 bits, il y a ... couleurs possibles.
```

Les `...` doivent être remplacés par le résultat de $2^{24}$, calculé par Python.

!!!


In [ ]:
# Exercice 12


## 🔡 Le code ASCII

En Python, il est possible de connaître la **position d’un caractère** dans la table ASCII grâce à la fonction `ord(...)`.  

📌 **Exemple :**

```python
print(ord("A"))
```

Par exemple, exécuter le programme ci-dessous afin de voir...

In [ ]:
print(ord("A"))

...que le caractère `A` se trouve à la position 65.


!!! note &nbsp; Exercice 13

Écrire un programme Python qui affiche les **codes ASCII** des caractères de la chaîne suivante : `"Info"`

!!!


In [ ]:
# Exercice 13


En Python, il est possible d’obtenir un **caractère** à partir de sa **position dans la table ASCII** grâce à la fonction `chr(...)`. Par exemple, exécuter le programme ci-dessous afin de voir 


In [3]:
print(chr(77),chr(65),chr(84), chr(72), chr(83))

M A T H S


que les caractères aux positions 77 65 84 72 83 sont `"M"` `"A"` `"T"` `"H"` `"S"`


## Du binaire directement au caractère

On peut combiner les deux idées :

1. `int("01000001", 2)` transforme le binaire en entier ;
2. `chr(...)` transforme cet entier en caractère.

In [4]:
# binaire --> entier
print(int("01000001", 2))

# En une seul étape: binaire --> entier --> caractère
print(chr(int("01000001", 2)))

65
A


!!! note &nbsp; Exercice 14: Message secret

Écrire un programme Python qui reconstitue le texte représenté par la suite de **codes binaires ASCII** suivante :  

`01000011 00100111 01100101 01110011 01110100 00100000 01100110 01100001 01100011 01101001 01101100 01100101`

!!!


In [ ]:
# Exercice 14

### L'ASCII Art

L’ASCII art est une forme d’art numérique qui utilise uniquement des caractères du code ASCII (lettres, chiffres, symboles) pour créer des images, motifs ou représentations visuelles.

```
       -''--.
       _`>   `\.-'<
    _.'     _     '._
  .'   _.='   '=._   '.
  >_   / /_\ /_\ \   _<
    / (  \o/\\o/  ) \
    >._\ .-,_)-. /_.<
        /__/ \__\ 
          '---'     E=mc^2
```


!!! note &nbsp; Exercice 15

Ecrire un programme qui affiche un sapin:
```
   X
  XXX
 XXXXX
XXXXXXX
   X
   X
```

!!!


In [ ]:
# Exercice 15



### Une astuce pour les dessins sur plusieurs lignes

Python permet d'écrire une chaîne de caractères sur plusieurs lignes avec trois guillemets :

```python
print("""
ligne 1
ligne 2
ligne 3
""")
```

Pour les dessins contenant beaucoup de `\`, on peut utiliser une **chaîne brute** en ajoutant `r` devant les trois guillemets.

!!! note &nbsp; Exercice 16

Ecrire un programme qui affiche les annimaux suivants
```
|\---/|
| o_o |
 \_^_/
```
```
       _.-- ,.--.
     .'   .'      /
     | @       |'..--------._
    /      \\._/              '.
   /  .-.-                     \\
  (  /    \\                     \\
  \\      '.                  | #
   \\       \\   -.           /
    :\\       |    )._____.'   
     "       |   /  \\  |  \\
             |   |./'  :__ \\
             '--'
        ELEPHANT
```
*Voir sur [www.asciiart.eu](https://www.asciiart.eu/animals) d'autres idées*

!!!


In [7]:
# à compléter



# Afficher des caractères Unicode

Python permet d'insérer un caractère à partir de son code Unicode directement dans une chaîne.

Deux écritures sont utiles :

- `\uXXXX` : exactement **4 chiffres hexadécimaux** ;
- `\UXXXXXXXX` : exactement **8 chiffres hexadécimaux**.

Par exemple :

In [8]:
print("\u27A9")
print("\U0001F680")

➩
🚀


!!! note &nbsp; Exercice 17

Cherchez quelques caractères Unicode qui vous intéressent et affichez-les avec Python.

Quelques idées :

- symboles mathématiques ;
- flèches ;
- monnaies ;
- alphabets ;
- pictogrammes ;
- emoji.

Vous pouvez utiliser la page Wikipédia  
[List of Unicode characters](https://en.wikipedia.org/wiki/List_of_Unicode_characters).

!!!


In [ ]:
# Exercice 17



!!! warning Checkpoint — Est-ce que j'ai compris ?

Essayez de répondre sans exécuter Python.

1. Quelle différence y a-t-il entre `print("3+3")` et `print(3+3)` ?
2. Que signifie l'opérateur `//` ?
3. Que signifie l'opérateur `%` ?
4. Comment reconnaître un nombre pair avec `%` ?
5. Que vaut `10110₂` en décimal ?
6. Combien de combinaisons peut-on coder avec 8 bits ?
7. Pourquoi `0.1 + 0.2` n'est-il pas exactement égal à `0.3` en Python ?
8. Quel lien existe entre un caractère et un nombre dans ASCII/Unicode ? 
!!!

---

#### Remarque générale

Ce document a été élaboré par les enseignant du collège Sismondi à Genève. Il est sous license Creative Commons [BY-NC-SA](https://creativecommons.org/licenses/?lang=fr)
![Licence Creative Commons](https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png)
